### PR TITLE
[ESIMD] Change simd::operator[] to return simd_view

### PR DESCRIPTION
**Pull request by DenisBakhvalov**
I'm working on the E2E test.

### CODE REVIEW

**Comment by DenisBakhvalov** [comment]
> I'm working on the E2E test.

The test is here: https://github.com/intel/llvm-test-suite/pull/323

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 1]
```diff
@@ -459,6 +437,124 @@ template <typename BaseTy, typename RegionTy> class simd_view {
   RegionTy M_region;
 };
 
+/// The reference class.
+///
+/// This class represents a region applied to a base object, which
+/// must be a simd object.
+///
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:444]:
```suggestion
/// This class represents a reference to a sub-region of a base simd object.
/// The referenced sub-region of the base object can be read from and written to via
/// an an instance of this class representing the reference.
///
```

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 2]
```diff
@@ -53,14 +53,16 @@ template <typename BaseTy, typename RegionTy> class simd_view {
...
-  simd_view(BaseTy &Base, RegionTy Region) : M_base(Base), M_region(Region) {}
-  simd_view(BaseTy &&Base, RegionTy Region) : M_base(Base), M_region(Region) {}
+protected:
+  simd_view_impl(BaseTy &Base, RegionTy Region)
+      : M_base(Base), M_region(Region) {}
+  simd_view_impl(BaseTy &&Base, RegionTy Region)
+      : M_base(Base), M_region(Region) {}
 
 public:
   // Disallow copy and move constructors for simd_view.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:63]:
```suggestion
  // Disallow copy and move constructors for simd_view_impl.
```

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 3]
```diff
@@ -20,15 +20,15 @@ namespace intel {
...
 
-/// The reference class.
+/// The simd_view base class.
 ///
 /// This class represents a region applied to a base object, which
 /// must be a simd object.
 ///
 /// \ingroup sycl_esimd
-template <typename BaseTy, typename RegionTy> class simd_view {
+template <typename BaseTy, typename RegionTy> class simd_view_impl {
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:29]: this class should be moved to the detail namespace

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 4]
```diff
@@ -20,15 +20,15 @@ namespace intel {
 namespace experimental {
 namespace esimd {
 
-/// The reference class.
+/// The simd_view base class.
 ///
 /// This class represents a region applied to a base object, which
 /// must be a simd object.
 ///
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:27]:
```suggestion
/// The simd_view base class.
///
/// Internal class implementing basic functionality of simd_view.
///
```

Nit: also would be good to add description why the split is needed.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp [hunk 5]
```diff
@@ -1292,6 +1292,23 @@ esimd_cbit(T src) {
   return Result[0];
 }
 
+/// Scalar version of \c esimd_cbit.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1295]: Please reformulate. The above function say is it is 'Scalar version of \c esimd_cbit' too. Maybe add comment in which cases this one is invoked.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 6]
```diff
@@ -459,6 +437,124 @@ template <typename BaseTy, typename RegionTy> class simd_view {
...
+  DEF_RELOP(>)
+  DEF_RELOP(>=)
+  DEF_RELOP(<)
+  DEF_RELOP(<=)
+  DEF_RELOP(==)
+  DEF_RELOP(!=)
+};
+
+template <bool Is2D, typename T, int StrideY, int StrideX>
+using region_base_1 = region_base<Is2D, T, 1, StrideY, 1, StrideX>;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:507]: to ::detail?

**Review by kbobrovs — COMMENTED** [review]
@rolandschulz, please review as well

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 3 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:29]: The `simd_view_impl` class defines all the public interfaces of `simd_view` class. Do we really want to hide it?

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 6 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:507]: Right. Will do.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 3 above)
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:29]: yes, because users are not supposed to ever use simd_view_impl

Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:29]: Right. Then I think we should also move simd_view_impl class into a separate header file, e.g. `detail/simd_view_impl.hpp`.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 2 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:63]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 1 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:444]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 4 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:27]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp (hunk 5 above)
Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1295]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp (hunk 3 above)
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp]: sounds good

Comment (DenisBakhvalov) [reply sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp]: Done.

**Review by rolandschulz — APPROVED** [review]
LGTM (Didn't review everything in detail)

File: sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp [hunk 7]
```diff
@@ -20,219 +19,45 @@ namespace intel {
...
+/// The referenced sub-region of the base object can be read from and written to
+/// via an instance of this class.
 ///
 /// \ingroup sycl_esimd
-template <typename BaseTy, typename RegionTy> class simd_view {
+template <typename BaseTy, typename RegionTy>
+class simd_view : public detail::simd_view_impl<BaseTy, RegionTy> {
   template <typename, int> friend class simd;
-  template <typename, typename> friend class simd_view;
+  // template <typename, typename> friend class simd_view;
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/simd_view.hpp:30]: Nit
```suggestion
```
please clean-up later

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**